src/components/CodeLineShapes/index.tsx

```diff
@@ -0,0 +1,39 @@
+import * as React from 'react';
+
+import { LineShapes, Token } from './utils';
+import styles from './styles.module.scss';
+
+export type PublicProps = {
+  lineShapes: LineShapes;
+};
+
+type Props = PublicProps;
+
+const CodeLineShapes = ({ lineShapes }: Props) => {
+  return (
+    <React.Fragment>
+      {lineShapes.tokens.map((shape) => {
+        let className;
+        if (shape.token === Token.whitespace) {
+          className = styles.whitespace;
+        } else {
+          className = styles.code;
+        }
+
+        return (
+          <div
+            key={[
+              String(lineShapes.line),
+              String(shape.percentOfWidth),
+              className,
+            ].join(':')}
+            className={className}
+            style={{ width: `${shape.percentOfWidth}%` }}
```

Comment:
I am still a bit worried that this will be a problem with CSP but maybe not.. We'll see soon.